File: net/src/main/java/com/topjohnwu/net/ProgressInputStream.java
package com.topjohnwu.net;

import java.io.FilterInputStream;
import java.io.IOException;
import java.io.InputStream;

public class ProgressInputStream extends FilterInputStream {

    private long totalBytes;
    private long bytesDownloaded;
    private DownloadProgressListener progress;

    public ProgressInputStream(InputStream in, long total, DownloadProgressListener listener) {
        super(in);
        totalBytes = total;
        progress = listener;
    }

    @Override
    public int read() throws IOException {
        int b = super.read();
        if (totalBytes > 0 && b >= 0) {
            bytesDownloaded++;
            Networking.mainHandler.post(() -> progress.onProgress(bytesDownloaded, totalBytes));
        }
        return b;
    }

    @Override
    public int read(byte[] b) throws IOException {
        return read(b, 0, b.length);
    }

    @Override
    public int read(byte[] b, int off, int len) throws IOException {
        int sz = super.read(b, off, len);
        if (totalBytes > 0 && sz > 0) {
            bytesDownloaded += sz;
            Networking.mainHandler.post(() -> progress.onProgress(bytesDownloaded, totalBytes));
        }
        return sz;
    }
}

File: app/core/src/main/java/com/topjohnwu/magisk/core/utils/ProgressInputStream.kt
package com.topjohnwu.magisk.core.utils

import java.io.FilterInputStream
import java.io.InputStream

class ProgressInputStream(
    base: InputStream,
    val progressEmitter: (Long) -> Unit
) : FilterInputStream(base) {

    private var bytesRead = 0L
    private var lastUpdate = 0L

    private fun emitProgress() {
        val cur = System.currentTimeMillis()
        if (cur - lastUpdate > 1000) {
            lastUpdate = cur
            progressEmitter(bytesRead)
        }
    }

    override fun read(): Int {
        val b = read()
        if (b >= 0) {
            bytesRead++
            emitProgress()
        }
        return b
    }

    override fun read(b: ByteArray): Int {
        return read(b, 0, b.size)
    }

    override fun read(b: ByteArray, off: Int, len: Int): Int {
        val sz = super.read(b, off, len)
        if (sz > 0) {
            bytesRead += sz
            emitProgress()
        }
        return sz
    }

    override fun close() {
        super.close()
        progressEmitter(bytesRead)
    }
}
